clients/kafka-common/src/main/java/org/eclipse/hono/kafka/client/CachingKafkaProducerFactory.java

```diff
@@ -0,0 +1,171 @@
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.BiFunction;
+
+import org.apache.kafka.common.errors.AuthorizationException;
+import org.apache.kafka.common.errors.OutOfOrderSequenceException;
+import org.apache.kafka.common.errors.ProducerFencedException;
+import org.apache.kafka.common.errors.UnsupportedForMessageFormatException;
+import org.apache.kafka.common.errors.UnsupportedVersionException;
+import org.eclipse.hono.kafka.client.test.FakeProducer;
+
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import io.vertx.core.Vertx;
+import io.vertx.kafka.client.producer.KafkaProducer;
+
+/**
+ * A factory for creating Kafka producers.
+ * <p>
+ * This implementation provides no synchronization and should not be used by multiple threads.
+ * <p>
+ * Created producers are being cached.
+ * <p>
+ * Producers are closed and removed from the cache if they throw a {@link #isFatalError(Throwable) fatal exception}. A
+ * following invocation of {@link #getOrCreateProducer(String, Map)} will then return a new instance.
+ *
+ * @param <K> The type for the record key serialization.
+ * @param <V> The type for the record value serialization.
+ */
+public class CachingKafkaProducerFactory<K, V> {
+
+    private final Map<String, KafkaProducer<K, V>> activeProducers = new HashMap<>();
+    private final BiFunction<String, Map<String, String>, KafkaProducer<K, V>> producerInstanceSupplier;
+
+    /**
+     * Creates a new producer factory.
+     *
+     * @param producerInstanceSupplier The function that provides new producer instances.
+     */
+    private CachingKafkaProducerFactory(
+            final BiFunction<String, Map<String, String>, KafkaProducer<K, V>> producerInstanceSupplier) {
+        this.producerInstanceSupplier = producerInstanceSupplier;
+    }
+
+    /**
+     * Creates an instance of the factory which produces {@link KafkaProducer#createShared(Vertx, String, Map) shared
+     * producers}.
+     * <p>
+     * Config must always be the same for the same key in {@link #getOrCreateProducer(String, Map)}.
+     *
+     * @param vertx The Vert.x instance to use.
+     * @param <K> The type for the record key serialization.
+     * @param <V> The type for the record value serialization.
+     * @return An instance of the factory.
+     */
+    public static <K, V> CachingKafkaProducerFactory<K, V> sharedProducerFactory(final Vertx vertx) {
+        return new CachingKafkaProducerFactory<>((name, config) -> KafkaProducer.createShared(vertx, name, config));
+    }
+
+    /**
+     * Creates an instance of the factory which produces {@link FakeProducer}s.
+     * <p>
+     * This is intended for tests only.
+     *
+     * @param <K> The type for the record key serialization.
+     * @param <V> The type for the record value serialization.
+     * @return An instance of the factory.
+     */
+    public static <K, V> CachingKafkaProducerFactory<K, V> testProducerFactory() {
```

Comment:
If we want to be able to create fake producers then I guess we should have a *KafkaProducerFactory* interface which can be mocked for test cases ...